REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [116]:
import pandas as pd

In [117]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [118]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [119]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [120]:
len(bus)

6315

#### Question 2b: How many unique business IDs are there?  

In [121]:
bus['business_id'].nunique()

6315

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [122]:
bus['name'].value_counts().head(5)

name
STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: count, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [123]:
type(bus['postal_code'].iloc[1])

str

#### Question 3b

What are the unique values of postal_code?

In [124]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [125]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [126]:
vb = bus[bus['postal_code'].isin(validZip)]

In [127]:
bus_valid = bus[bus['postal_code'].isin(validZip)]
len(bus_valid)

5999

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [128]:
vb[pd.isnull(vb['longitude'])]

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

In [129]:
sum(bus_valid['longitude'].isnull())

2483

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [130]:
new_df = pd.DataFrame({'postal_code': validZip})
new_df['null_lon'] = 0
new_df['not_null_lon'] = 0


for zip in validZip:
    filtered = bus[bus["postal_code"] == zip]
    new_df.loc[new_df['postal_code'] == zip, 'null_lon'] = filtered['longitude'].isnull().sum()
    new_df.loc[new_df['postal_code'] == zip, 'not_null_lon'] = filtered['longitude'].notnull().sum()

print(new_df)

   postal_code  null_lon  not_null_lon
0        94102       196           262
1        94103       252           306
2        94104        59            74
3        94105       100           126
4        94107       247           204
5        94108        82           142
6        94109       146           237
7        94110       251           334
8        94111       115           167
9        94112        59           131
10       94114       108           122
11       94115        76           162
12       94116        35            58
13       94117        67           130
14       94118       100           143
15       94121        55           106
16       94122       123           147
17       94123        53           116
18       94124        88            76
19       94127        28            43
20       94131        14            34
21       94132        65            69
22       94133       135           282
23       94134        29            45


#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [131]:
new_df['prop_null_lon'] = new_df['null_lon']/(new_df['null_lon'] + new_df['not_null_lon'])
new_df_2 = pd.DataFrame({
   'postal_code': new_df['postal_code'], 
    'prop_null_lon': new_df['prop_null_lon'] 
})
new_df_2.sort_values(by = "prop_null_lon", ascending=False).head(5)



postal_code  prop_null_lon
4        94107       0.547672
18       94124       0.536585
21       94132       0.485075
10       94114       0.469565
16       94122       0.455556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [132]:
print(len(ins))
print(ins['business_id'].nunique())

len(ins) == ins['business_id'].nunique()

15430
5730


False

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [133]:
type(ins['type'].iloc[1])

str

In [138]:
ins2 = ins[ins['type'].map(ins['type'].value_counts()) > 10]

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [140]:
ins2['date'] = ins2['date'].astype(str)
ins2.loc[:, 'year'] = ins2['date'].str[:4].astype(int)
ins2

C:\Users\aaa\AppData\Local\Temp\ipykernel_8216\1642654837.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2['date'] = ins2['date'].astype(str)
C:\Users\aaa\AppData\Local\Temp\ipykernel_8216\1642654837.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2.loc[:, 'year'] = ins2['date'].str[:4].astype(int)


business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15429 rows x 5 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [141]:
print(ins2['year'].unique())
print(ins2['year'].value_counts())

[2016 2014 2015 2013]
year
2016    5839
2014    5629
2015    3923
2013      38
Name: count, dtype: int64


In [142]:
ins3 = ins2.groupby('year').filter(lambda x: len(x) >= 50)
ins3

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15391 rows x 5 columns]

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [143]:
ins3[ins3['year'] == 2016]

business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5839 rows x 5 columns]

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [150]:
bus_ins = pd.merge(bus, ins3, on = 'business_id', how="left")

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [152]:
bottom_20 = bus_ins.sort_values(by="score").head(20)
bottom_20[['name', 'address', 'score']]

name                     address  \
10664                        ABC BAKERY CAFE             650 JACKSON ST    
6362                         IMPERIAL PALACE          818 WASHINGTON ST    
9258          YUMMY DIM SUM & FAST FOOD, LLC            930 STOCKTON ST    
11708                GOLDEN RIVER RESTAURANT            5827 GEARY BLVD    
2446                             LES JOULINS                44 ELLIS ST    
2426                        YEE'S RESTAURANT             1131 GRANT AVE    
6710                     BRISTOL FARMS #2317           845 MARKET ST 010   
5907                           BODEGA BISTRO              607 LARKIN ST    
8629        OZONE THAI RESTAURANT AND LOUNGE                598 02ND ST    
2715             BASIL THAI RESTAURANT & BAR             1175 FOLSOM ST    
12076                        MISS SAIGON LLC                100 06TH ST    
11665                             UNCLE CAFE              65 WAVERLY PL    
14587                HARBOR VILLA RESTAURANT  5238 DIAMOND HEIGHTS BLVD    
7749                              FIRST CAKE            5901 GEARY BLVD    
15689                              POKI TIME            2101 LOMBARD ST    
14455                           CRAZY PEPPER          2257 SAN JOSE AVE    
11651                          SA BEANG THAI          312 DIVISADERO ST    
3362                         GOURMET KITCHEN           1051 STOCKTON ST    
14400  LE SOLEIL INTERNATIONAL HOLDINGS, INC          133 CLEMENT  ST.     
858                          TAQUERIA LA PAZ                305 TURK ST    

       score  
10664   46.0  
6362    47.0  
9258    50.0  
11708   52.0  
2446    53.0  
2426    53.0  
6710    54.0  
5907    54.0  
8629    54.0  
2715    54.0  
12076   54.0  
11665   55.0  
14587   55.0  
7749    55.0  
15689   55.0  
14455   55.0  
11651   55.0  
3362    55.0  
14400   55.0  
858     56.0

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!